pkg/localstorev2/internal/chunkstore/chunkstore.go

```diff
@@ -0,0 +1,338 @@
+	}
+	if c.Stamp == nil {
+		return nil, errMarshalInvalidChunkStampItemStamp
+	}
+	buf, err := c.Stamp.MarshalBinary()
+	if err != nil {
+		return nil, fmt.Errorf("failed marshaling stamp: %w", err)
+	}
+	return buf, nil
+}
+
+func (c *chunkStampItem) Unmarshal(buf []byte) error {
+	// ensure that the address is set already in the item
+	if c.Address.IsZero() {
+		return errUnmarshalInvalidChunkStampItemAddress
+	}
+	stamp := new(postage.Stamp)
+	if err := stamp.UnmarshalBinary(buf); err != nil {
+		if errors.Is(err, postage.ErrStampInvalid) {
+			return errInvalidChunkStampItemSize
+		}
+		return fmt.Errorf("failed unmarshaling stamp: %w", err)
+	}
+
+	ni := new(chunkStampItem)
+	ni.Address = swarm.NewAddress(append(make([]byte, 0, swarm.HashSize), c.Address.Bytes()...))
+	ni.Stamp = stamp
+	*c = *ni
+	return nil
+}
+
+type chunkStoreWrapper struct {
+	st     storage.Store
+	sharky *sharky.Store
+}
+
+// New returns a chunkStoreWrapper which implements the storage.ChunkStore interface
+// on top of the Store and the sharky blob storage.
+// It uses the Store to maintain the retrievalIndex and also stamps for the chunk and
+// sharky to maintain the chunk data.
+// Currently we do not support multiple stamps on chunks, but the chunkStoreWrapper
+// allows for this in future.
+// chunkStoreWrapper uses refCnt and works slightly different to the ChunkStore API.
+// The current Put API returns true if the chunk is already found as it supports only
+// unique chunks. The chunkStoreWrapper will also maintain refCnts for all the Puts
+// for the same chunk. This allows duplicates to change state.
+// Usually this Put operation would not be used by clients directly as it will be part
+// of some component which will provide the uniqueness guarantee.
+// Due to the refCnt a Delete would only result in an actual Delete operation
+// if the refCnt goes to 0.
+func New(st storage.Store, sharky *sharky.Store) storage.ChunkStore {
+	return &chunkStoreWrapper{st: st, sharky: sharky}
+}
+
+func (c *chunkStoreWrapper) getStamp(addr swarm.Address) (swarm.Stamp, error) {
+	var stamp swarm.Stamp
+	err := c.st.Iterate(storage.Query{
+		Factory: func() storage.Item { return &chunkStampItem{Address: addr} },
+	}, func(res storage.Result) (bool, error) {
+		// Use the first available stamp till we support multiple stamps on chunk
+		stamp = res.Entry.(*chunkStampItem).Stamp
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return stamp, nil
+}
+
+func (c *chunkStoreWrapper) removeStamps(addr swarm.Address) error {
+	var stamps []swarm.Stamp
+	err := c.st.Iterate(storage.Query{
+		Factory: func() storage.Item { return &chunkStampItem{Address: addr} },
+	}, func(res storage.Result) (bool, error) {
+		stamps = append(stamps, res.Entry.(*chunkStampItem).Stamp)
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+	var delErr error
+	for _, s := range stamps {
+		err := c.st.Delete(&chunkStampItem{Address: addr, Stamp: s})
```

Comment:
So the problem here is around the locking. In our inmem implementation, each update operation needs an exclusive lock. This is why a Delete during Iteration is not possible. Although this would be possible on levelDB. I think this is generally a better pattern to delete as it would work well in both cases.

Also, chunk -> stamps will not be a lot many entries where this would cause us any problems.